persistence/sql/persister_verification.go

```diff
@@ -132,3 +136,152 @@
 	}
 	return nil
 }
+func (p *Persister) UseVerificationCode(ctx context.Context, fID uuid.UUID, codeVal string) (*code.VerificationCode, error) {
+	ctx, span := p.r.Tracer(ctx).Tracer().Start(ctx, "persistence.sql.UseVerificationCode")
+	defer span.End()
+
+	var verificationCode *code.VerificationCode
+
+	nid := p.NetworkID(ctx)
+
+	flowTableName := new(verification.Flow).TableName(ctx)
+
+	if err := sqlcon.HandleError(p.Transaction(ctx, func(ctx context.Context, tx *pop.Connection) (err error) {
+
+		/* #nosec G201 TableName is static */
+		if err := sqlcon.HandleError(
+			tx.RawQuery(
+				fmt.Sprintf("UPDATE %s SET submit_count = submit_count + 1 WHERE id = ? AND nid = ?", flowTableName),
+				fID,
+				nid,
+			).Exec(),
+		); err != nil {
+			return err
+		}
+
+		var submitCount int
+		// Because MySQL does not support "RETURNING" clauses, but we need the updated `submit_count` later on.
+		/* #nosec G201 TableName is static */
+		if err := sqlcon.HandleError(
+			tx.RawQuery(
+				fmt.Sprintf("SELECT submit_count FROM %s WHERE id = ? AND nid = ?", flowTableName),
+				fID,
+				nid).
+				First(&submitCount),
+		); err != nil {
+			if errors.Is(err, sqlcon.ErrNoRows) {
+				// Return no error, as that would roll back the transaction
+				return nil
+			}
+
+			return err
+		}
+		// This check prevents parallel brute force attacks to generate the verification code
+		// by checking the submit count inside this database transaction.
+		// If the flow has been submitted more than 5 times, the transaction is aborted (regardless of whether the code was correct or not)
+		// and we thus give no indication whether the supplied code was correct or not. See also https://github.com/ory/kratos/pull/2645#discussion_r984732899
+		if submitCount > 5 {
+			return errors.WithStack(code.ErrCodeSubmittedTooOften)
+		}
+
+		var verificationCodes []code.VerificationCode
+		if err = sqlcon.HandleError(
+			tx.Where("nid = ? AND selfservice_verification_flow_id = ?", nid, fID).
+				All(&verificationCodes),
+		); err != nil {
+			if errors.Is(err, sqlcon.ErrNoRows) {
+				// Return no error, as that would roll back the transaction
+				return nil
+			}
+
+			return err
+		}
+
+	secrets:
+		for _, secret := range p.r.Config().SecretsSession(ctx) {
+			suppliedCode := []byte(p.hmacValueWithSecret(ctx, codeVal, secret))
+			for i := range verificationCodes {
+				code := verificationCodes[i]
+				if subtle.ConstantTimeCompare([]byte(code.CodeHMAC), suppliedCode) == 0 {
+					// Not the supplied code
+					continue
+				}
+				verificationCode = &code
+				break secrets
```

Comment:
Literally had to google what this does. Epic, didn't know!

Reply:
One of the _very_ few situations where I used that in actual code. Nearly every language I wrote code in supports it, but this was the first situation that actually required it :D